### DISCUSSION TITLE
it seems too slow

### DISCUSSION THREAD

**Discussion by ghost** [gfx-rs/wgpu#3128] [Q&A] [answered]
for the triangle tutorial:
wgpu-rs - 500fps
vulkano - 1000fps
how to improve performance

**Comment by grovesNL** [comment, marked as answer]
Hi @zzwxh! :wave: At this framerate, the difference between 500fps and 1000fps probably isn't meaningful, because it's sensitive to very small differences between both examples.

There are lots of things that could contribute to this difference, for example:
- Overhead of wgpu: wgpu has some amount of non-zero overhead for some things like resource tracking, but for most use cases (e.g., realistic scenes) this shouldn't be significant. In other words, even if the overhead increases the frame time of one triangle being rendered like this example, we'd generally expect this overhead to be minimal compared to a real application that has implements its own resource tracking anyway (e.g., an application running vulkano or ash).
- Different window sizes: if the window size is larger, the program may be slower to run (i.e., limited by fill rate), which could account for differences between your wgpu wgpu and vulkano examples if the window size is larger for wgpu.
- Different shaders: there are lots of ways to write a simple triangle shader, and at these kind of frame times, any difference in the shader can cause noticeable differences (e.g., indexing vertices vs. bit swizzling a vertex index). The shaders should be the exact same.
- Using a different graphics adapter: if there are multiple graphics adapters, you could be running wgpu and vulkano on different graphics adapters depending on how the graphics adapter is selected.
- Debug vs. release mode: benchmarks around frame time like this one should only use release mode.

Hope this helps! Overall I'd recommend not worrying too much about the performance of a single triangle. Rendering a single triangle at 500fps vs. 1000fps vs. 2000fps isn't a great indicator of how the crates will perform. Instead the performance really matters once you start testing these crates with realistic workloads, and generally those paths will be the most optimized.
